plugin-testing-framework/script/platform.ts

```diff
@@ -0,0 +1,512 @@
+    }
+    
+    /**
+     * Restarts an already installed application by app id.
+     */
+    restartApplication(appId: string): Q.Promise<string> {
+        return this.endRunningApplication(appId)
+            .then<void>(() => {
+                // Wait for a second before restarting.
+                return Q.delay(1000);
+            })
+            .then<string>(() => {
+            return this.launchInstalledApplication(appId);
+        });
+    }
+    
+    /**
+     * Navigates away from the current app, waits for a delay (defaults to 1 second), then navigates to the specified app.
+     */
+    resumeApplication(appId: string, delayBeforeResumingMs: number = 1000): Q.Promise<string> {
+        // Open a default Android app (for example, settings).
+        return this.launchInstalledApplication("com.android.settings")
+            .then<void>(() => {
+                console.log("Waiting for " + delayBeforeResumingMs + "ms before resuming the test application.");
+                return Q.delay(delayBeforeResumingMs);
+            })
+            .then<string>(() => {
+                // Reopen the app.
+                return this.launchInstalledApplication(appId);
+            });
+    }
+    
+    /**
+     * Prepares the emulator for a test.
+     */
+    prepareEmulatorForTest(appId: string): Q.Promise<string> {
+        return this.endRunningApplication(appId)
+            .then(() => { return tu.TestUtil.getProcessOutput("adb shell pm clear " + appId); });
+    }
+    
+    /**
+     * Uninstalls the app from the emulator.
+     */
+    uninstallApplication(appId: string): Q.Promise<string> {
+        return tu.TestUtil.getProcessOutput("adb uninstall " + appId);
+    }
+}
+
+export class IOSEmulatorManager implements IEmulatorManager {
+    
+    private static IOS_EMULATOR_OPTION_NAME: string = "--iosemu";
+    
+    private targetEmulator: string;
+    
+    /**
+     * Returns the target emulator, which is specified through the command line.
+     */
+    getTargetEmulator(): Q.Promise<string> {
+        if (this.targetEmulator) return Q<string>(this.targetEmulator);
+        else {
+            var deferred = Q.defer<string>();
+        
+            var targetIOSEmulator: string = tu.TestUtil.readMochaCommandLineOption(IOSEmulatorManager.IOS_EMULATOR_OPTION_NAME);
+            
+            if (!targetIOSEmulator) {
+                // If no iOS simulator is specified, get the most recent iOS simulator to run tests on.
+                tu.TestUtil.getProcessOutput("xcrun simctl list", { noLogCommand: true, noLogStdOut: true, noLogStdErr: true })
+                    .then<string>(
+                        (listOfDevices: string) => {
+                            var phoneDevice = /iPhone (\S* )*(\(([0-9A-Z-]*)\))/g;
+                            var match = listOfDevices.match(phoneDevice);
+                            deferred.resolve(match[match.length - 1]);
+                            return undefined;
```

Comment:
i used `Q.Promise<string>` everywhere when i built this code, mostly because i was having trouble with the typing and being consistent with one type seemed to resolve it (i was new to promises at the time). i used `string` specifically because the only time these promise chains return anything defined is in the TestUtil.getProcessOutput function. i could probably go through and fix it all if you think that would be significantly better.